pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/multifile/ApexMultifileAnalysis.java

```diff
@@ -0,0 +1,96 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.apex.multifile;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Logger;
+
+import com.nawforce.common.api.FileIssueOptions;
+import com.nawforce.common.api.Org;
+import com.nawforce.common.api.ServerOps;
+import com.nawforce.common.diagnostics.Issue;
+
+/**
+ * Stores multi-file analysis data. The 'Org' here is the primary ApexLink structure for maintaining information
+ * about the Salesforce metadata. We load 'Packages' into it to perform analysis. Once constructed you
+ * can get 'Issue' information from it on what was found. The 'Org' holds mutable state for IDE use that can get quite
+ * large (a few hundred MB on very large projects). An alternative way to use this would be to cache the
+ * issues after packages are loaded and throw away the 'Org'. That would be a better model if all you wanted was the
+ * issues but more complex rules will need the ability to traverse the internal graph of the 'Org'.
+ *
+ * @author Kevin Jones
+ */
+@SuppressWarnings("PMD")
+public final class ApexMultifileAnalysis {
+    private static final Logger LOG = Logger.getLogger(ApexMultifileAnalysis.class.getName());
+
+    private static Map<String, ApexMultifileAnalysis> instanceMap = new HashMap<>();
+
+    // An arbitrary large number of errors to report
+    private static final Integer MAX_ERRORS_PER_FILE = 100;
+
+    // Create a new org for each analysis
+    private Org org = Org.newOrg(true);
+    private FileIssueOptions options = new FileIssueOptions();
+
+    private ApexMultifileAnalysis(String multiFileAnalysisDirectory) {
+        LOG.fine("MultiFile Analysis created for " + multiFileAnalysisDirectory);
+        if (multiFileAnalysisDirectory != null && !multiFileAnalysisDirectory.isEmpty()) {
+            // Default issue options, zombies gets us unused methods & fields as well as deploy problems
+            options.includeZombies_$eq(true);
+            options.maxErrorsPerFile_$eq(MAX_ERRORS_PER_FILE);
+
+            // Load the package into the org, this can take some time!
+            org.newSFDXPackage(multiFileAnalysisDirectory);
+            org.flush();
+        }
+    }
+
+    public Issue[] getFileIssues(String filename) {
```

Comment:
Please return an unmodifiable `List` here (or even an `Iterable<Issue>`). We rarely use arrays in the codebase. You can write `Collections.unmodifiableList(Arrays.asList(...))`

Reply:
Done (in ASTApexFile.getGlobalIssues() now)